Spring Kafka Project

obedjamir opened the pull request:
first commit

obedjamir committed:
Spring Kafka Project
first commit

PinakiGhosh commented on pom.xml:
```diff
@@ -0,0 +1,96 @@
+			<version>${opencsv.version}</version>
+		</dependency>
+		<dependency>
+			<groupId>ch.qos.logback</groupId>
```
This is not required

PinakiGhosh commented on service/src/main/java/com/bankbazaar/kafka/service/controller/KafkaController.java:
```diff
@@ -0,0 +1,25 @@
+
+@RestController
+public class KafkaController {
+    private final KafkaProducer producer;
```
Autowire the KafkaProducer

obedjamir committed:
Spring Kafka Project
Mapper errors resolved

obedjamir committed:
Test case implemented
Checks if file csv file is created
Timeout for file creation set to 10 sec
Contents of csv file are validated

PinakiGhosh commented on pom.xml:
```diff
@@ -49,31 +47,18 @@
 			<artifactId>spring-boot-starter-validation</artifactId>
 			<version> ${spring-boot-starter-validation.version}</version>
 		</dependency>
+		<dependency>
```
Move all dependencies to the service pom

obedjamir replied:
dependencies moved to service pom

PinakiGhosh commented on service/src/main/java/com/bankbazaar/kafka/service/consumer/KafkaConsumer.java:
```diff
@@ -13,7 +13,7 @@
 @Service
 public class KafkaConsumer {
-    @KafkaListener(topics = "create_CSV")
+    @KafkaListener(topics = "create_CSV", groupId = "Partner_Integration_Kafka_Group")
```
GroupId should be something related to your project. Also keep all configs like this in application.properties

obedjamir replied:
group id changed

PinakiGhosh commented on service/src/test/java/com/bankbazaar/kafka/service/controller/KafkaControllerTest.java:
```diff
@@ -0,0 +1,55 @@
+class KafkaControllerTest {
+
+    @Autowired
+    private KafkaController kafkaController;
```
Use MockMvc how we used in the previous project. Not like this

obedjamir replied:
Mockmvc implemented for testing

PinakiGhosh commented on service/src/test/java/com/bankbazaar/kafka/service/controller/KafkaControllerTest.java:
```diff
@@ -0,0 +1,55 @@
+        while ((nextRecord = csvReader.readNext()) != null)
+        {
+            assertArrayEquals(nextRecord, dataDto.getData()[i].split("\\|"));
+            i++;
```
Also assert on the number of lines

obedjamir replied:
this has been implemented

obedjamir committed:
Requested changes implemented
moved dependencies to service pom.xml
assert number of lines
used mockmvc for testing
groud id changed

obedjamir committed:
Kafka Stream Config Issues
Problems with starting up Kafka

obedjamir committed:
kafa streams implementation
added a kafka stream processor to validate file creation
the processor checks if files already exits
Also the status of file creation is saved in mysql DB

obedjamir committed:
TestCases Implemented
checks file creation and file content
checks status in mysql DB
checks if file already exists

obedjamir committed:
Removed unnecessary imports

PinakiGhosh commented on core/src/main/java/com/bankbazaar/kafka/core/model/Data.java:
```diff
@@ -8,6 +8,7 @@
 @NoArgsConstructor
 @AllArgsConstructor
 public class Data {
+    private Long id;
```
Why is an id required?

obedjamir replied:
I included id so that i can update the status in db after each process.

PinakiGhosh commented on dto/src/main/java/com/bankbazaar/kafka/dto/model/FileStatusDto.java:
```diff
@@ -0,0 +1,16 @@
+@lombok.Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class FileStatusDto {
```
No need for a FileStatusDto

obedjamir replied:
FileStatusDto removed

PinakiGhosh commented on service/pom.xml:
```diff
@@ -13,10 +13,25 @@
+        <kafka-streams.version>3.1.0</kafka-streams.version>
+        <awaitility.version>4.2.0</awaitility.version>
+        <spring.cloud.version>2021.0.2</spring.cloud.version>
+        <property name="hibernate.enable_lazy_load_no_trans" value="true"/>
```
Why is this required?

obedjamir replied:
I forgot to remove this. This is not a required line.

PinakiGhosh commented on service/src/main/java/com/bankbazaar/kafka/service/consumer/KafkaConsumer.java:
```diff
@@ -1,32 +1,72 @@
 
 @Slf4j
 @Service
 public class KafkaConsumer {
```
Rename to KafkaStreamsConfig. This class no longer is a single kafka consumer

obedjamir replied:
all KafkaStream bean moved to KafkaStreamsConfig

PinakiGhosh commented on service/src/main/java/com/bankbazaar/kafka/service/producer/KafkaProducer.java:
```diff
@@ -1,25 +1,41 @@
-    public void sendData(DataDto data)
+    public Response sendData(DataDto data)
     {
+        FileStatusDto fileData = new FileStatusDto();
```
Dtos are used in a scenario where we want to send data out of the application to a client. Its not required when we are calling other methods in the same service

obedjamir replied:
FileStatusDTo removed

PinakiGhosh commented on service/src/main/java/com/bankbazaar/kafka/service/producer/KafkaProducer.java:
```diff
@@ -1,25 +1,41 @@
+        FileStatusDto fileData = new FileStatusDto();
+        fileData.setFileName(data.getFileName());
+        fileData.setStatus("NEW");
+        FileStatusDto response = fileStatusService.insert(fileData);
```
This method should be like this
1. get Data object
2. Pass this object to the FileStatusService
3. FileStatusService should add the new entry and send back the entity
4. Push the request to the Kafka producer for processing
5. Create Response from the entity returned in step #3 and send it back

obedjamir replied:
This has been implemented

obedjamir committed:
Implemented Requested Changes
All kafka stream beans moved to KafkaStreamConfig
Removed FileStatusDto and FileStatusMapper
createEntry method added to FileStatusService

obedjamir committed:
Correction to annotatins
@Service changed to @Configuration in KafkaStreamConfig
Correction made in KafkaProducer

obedjamir committed:
Requested Changes Implemented
process complete notification on successful file cretion
KafaStreamConfig renamed to KafkaStreamsConfig
Unnecessary imports removed

obedjamir committed:
Correction made in KafkaStreamsConfig
log.error('process complete') changed to log.info('process complete')

PinakiGhosh commented on service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java:
```diff
@@ -0,0 +1,151 @@
+                                    if (file.exists()) {
+                                        fileData.setStatus(Status.FAILURE);
+                                        fileStatusService.update(fileData);
+                                        return false;
```
This method is supposed to return of type Data and you are returning a boolean.

obedjamir replied:
this method now return Data type

PinakiGhosh commented on service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java:
```diff
@@ -0,0 +1,151 @@
+    {
+        return kStream -> kStream.foreach((key, value) ->
+                {
+                    try {
```
The Retry template + context itself is like a try catch. No need to add it explicitly

obedjamir replied:
this has been removed

PinakiGhosh commented on service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java:
```diff
@@ -0,0 +1,151 @@
+                                        }
+                                        CsvWriter.close();
+                                        fileData.setStatus(Status.SUCCESS);
+                                        fileStatusService.update(fileData);
```
Post the initial creation we are only updating the status. I think it will be better to have a method updateStatus(Long id, Status status) for this purpose

obedjamir replied:
this has been implemented

PinakiGhosh commented on service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java:
```diff
@@ -0,0 +1,151 @@
+                                    }
+                                    return null;
+                                }, context -> {
+                                    log.error("retries exhausted");
```
You should log the actual exception as well

obedjamir replied:
this has been implemented

obedjamir committed:
Implemented Requested Changes
updateEntry method implemented in FileStatusService
log.error logs exception
try catch removed in consumer
